source/Halibut/Transport/Protocol/MessageExchangeStream.cs

```diff
@@ -246,14 +260,51 @@ void ExpectServerIdentity()
 
         T ReadBsonMessage<T>()
         {
-            using (var zip = new DeflateStream(stream, CompressionMode.Decompress, true))
-            using (var bson = new BsonDataReader(zip) { CloseInput = false })
+            using (var buffer = new BufferedReadStream(stream))
             {
-                var messageEnvelope = serializer.Deserialize<MessageEnvelope>(bson);
-                if (messageEnvelope == null)
-                    throw new Exception("messageEnvelope is null");
-                
-                return (T)messageEnvelope.Message;
+                try
+                {
+                    using (var zip = new DeflateStream(buffer, CompressionMode.Decompress, true))
+                    {
+                        using (var bson = new BsonDataReader(zip) {CloseInput = false})
+                        {
+                            var messageEnvelope = serializer.Deserialize<MessageEnvelope>(bson);
+
+                            log.Write(EventType.Diagnostic, $"ReadBsonMessage: {BitConverter.ToString(buffer.GetBytes())}");
+
+                            if (messageEnvelope == null)
+                                throw new Exception("messageEnvelope is null");
+                            return (T) messageEnvelope.Message;
+                        }
+                    }
+                }
+                catch (Exception ex)
+                {
+                    var plaintext = SafelyGetPlainText(buffer);
+                    if (plaintext.Equals(End))
+                        throw new HalibutClientException("Connection ended by remote.");
```

Comment:
I'm on the fence about re-using these exceptions. On one hand, we (appear to) get nice transient error handling automatically in Octopus. On the other hand, it might be nicer to be specific.
Either way, we want to:
a) suppress the stack trace
b) include an error code of some description to make it easy to track down where the error occured